ingest/src/server_iot.rs

```diff
@@ -35,38 +183,99 @@ impl ManagedTask for GrpcServer {
     ) -> LocalBoxFuture<'static, anyhow::Result<()>> {
         let address = self.address;
         Box::pin(async move {
-            transport::Server::builder()
+            let grpc_server = transport::Server::builder()
                 .layer(poc_metrics::request_layer!("ingest_server_iot_connection"))
                 .add_service(poc_lora::Server::new(*self))
-                .serve_with_shutdown(address, shutdown)
-                .map_err(Error::from)
-                .await
+                .serve(address)
+                .map_err(Error::from);
+
+            tokio::select! {
+                _ = shutdown => {
+                    tracing::warn!("grpc server is shutting down");
+                    Ok(())
+                }
+                _ = grpc_server => {
+                    Err(anyhow::anyhow!("grpc server stopping running"))
+                }
+            }
         })
     }
 }
 
-impl GrpcServer {
-    fn verify_network(&self, public_key: PublicKey) -> VerifyResult<PublicKey> {
-        if self.required_network == public_key.network {
-            Ok(public_key)
-        } else {
-            Err(Status::invalid_argument("invalid network"))
-        }
-    }
+fn verify_public_key(bytes: &[u8]) -> VerifyResult<PublicKey> {
+    PublicKey::try_from(bytes).map_err(|_| Status::invalid_argument("invalid public key"))
+}
 
-    fn verify_public_key(&self, bytes: &[u8]) -> VerifyResult<PublicKey> {
-        PublicKey::try_from(bytes).map_err(|_| Status::invalid_argument("invalid public key"))
+fn verify_network(required_network: Network, public_key: PublicKey) -> VerifyResult<PublicKey> {
+    if required_network == public_key.network {
+        Ok(public_key)
+    } else {
+        Err(Status::invalid_argument("invalid network"))
     }
+}
 
-    fn verify_signature<E>(&self, public_key: PublicKey, event: E) -> VerifyResult<(PublicKey, E)>
-    where
-        E: MsgVerify,
-    {
-        event
-            .verify(&public_key)
-            .map_err(|_| Status::invalid_argument("invalid signature"))?;
-        Ok((public_key, event))
-    }
+fn verify_signature<E>(public_key: Option<&PublicKey>, event: E) -> VerifyResult<E>
+where
+    E: MsgVerify,
+{
+    public_key
+        .ok_or_else(|| Status::invalid_argument("signing public key not available"))
+        .and_then(|public_key| {
+            event
+                .verify(public_key)
+                .map(|_| event)
+                .map_err(|_| Status::invalid_argument("invalid signature"))
+        })
+}
+
+async fn handle_beacon_report(
+    file_sink: &FileSinkClient,
+    timestamp: u64,
+    report: LoraBeaconReportReqV1,
+    signing_key: Option<&PublicKey>,
+    expected_pubkey_bytes: Option<&Vec<u8>>,
```

Comment:
can this be `Option<&[u8]>`? maybe with `bytes == report.pub_key.as_ref() | report.pub_key.as_slice()`?

Reply:
not sure why this is better since the pub_key comes as a `Vec<u8>` and we just store it to compare it to the same field in subsequent messages.

Reply:
Ok, Rondo convinced why this is better, will change

Reply:
Fixed ff72f1b